datasets/oscar/oscar.py

```diff
@@ -0,0 +1,369 @@
+        self.language = language
+        self.shuffled = shuffled
+        self.deduplicated = deduplicated
+        self.base_data_url = _BASE_DATA_URL_FORMAT_STR.format(
+            shuffled=shuffled_str, language=language, deduplicated=deduplicated_str
+        )
+
+
+class Oscar(datasets.GeneratorBasedBuilder):
+    """OSCAR The Open Super-large Crawled ALMAnaCH coRpus."""
+
+    # Version history:
+    # 1.0.0: S3 (new shuffling, sharding and slicing mechanism).
+    # 0.1.0: Initial version.
+    BUILDER_CONFIGS = [
+        OscarConfig(  # pylint: disable=g-complex-comprehension
+            language=language,
+            shuffled=False,
+            deduplicated=True,
+            version=datasets.Version("1.0.0"),
+        )
+        for language in _languages()
+    ] + [
+        OscarConfig(  # pylint: disable=g-complex-comprehension
+            language=language,
+            shuffled=False,
+            deduplicated=False,
+            version=datasets.Version("1.0.0"),
+        )
+        for language in _languages()
+    ]
+    BUILDER_CONFIG_CLASS = OscarConfig
+
+    def _info(self):
+        return datasets.DatasetInfo(
+            description=_DESCRIPTION,
+            features=datasets.Features({"id": datasets.Value("int64"), "text": datasets.Value("string")}),
+            supervised_keys=None,
+            homepage=_URL,
+            citation=_CITATION,
+            license=_LICENSE,
+        )
+
+    def _split_generators(self, dl_manager):
+        checksum_url = self.config.base_data_url + _BASE_CHECKSUM_FILE_NAME.format(language=self.config.language)
+        checksum_file = dl_manager.download(checksum_url)
+        with open(checksum_file, encoding="utf-8") as f:
+            data_filenames = [line.split("\t")[0] for line in f if line]
+            data_urls = [self.config.base_data_url + data_filename for data_filename in data_filenames]
+        downloaded_files = dl_manager.download(data_urls)
+        return [
+            datasets.SplitGenerator(name=datasets.Split.TRAIN, gen_kwargs={"filepaths": downloaded_files}),
+        ]
+
+    def _generate_examples(self, filepaths):
+        """This function returns the examples in the raw (text) form."""
+        id_ = 0
+        current_lines = []
+        for filepath in filepaths:
+            logging.info("generating examples from = %s", filepath)
+            with gzip.open(filepath, "rt") as f:
+                for line in f:
+                    if len(line.strip()) > 0:
+                        current_lines.append(line)
+                    else:
+                        feature = id_, {"id": id_, "text": "".join(current_lines).rstrip()}
+                        yield feature
+                        id_ += 1
+                        current_lines = []
+                # last paragraph
+                if current_lines:
+                    feature = id_, {"id": id_, "text": "".join(current_lines).rstrip()}
+                    yield feature
```

Comment:
Hi @lhoestq I think the `id_` needs to be incremented in the last paragraph condition too 🤔

Reply:
Yes you're right, good catch